community/cypher/frontend-3.3/src/main/scala/org/neo4j/cypher/internal/frontend/v3_3/ast/GraphReturnItems.scala

```diff
@@ -99,7 +99,8 @@ final case class GraphReturnItems(includeExisting: Boolean, items: Seq[GraphRetu
           newSourceName.map(source => newTargetName.map(target => ContextGraphs(source, target)).getOrElse(ContextGraphs(source, source)))
         optNewContextGraphs match {
           case Some(newContextGraphs) => s.updateContextGraphs(newContextGraphs)
-          case None => Left(SemanticError("No context graphs available", position))
+            // If there are no context graphs that's fine -- use the ambient graph
```

Comment:
But we want this to be an error if it happens regularly - it would only be ok for the leading with, wouldn't it? Or are you suggesting a change in semantics - no context graph specified, back to ambient?

Reply:
BTW we need to think this though for UNION, too, I think.

Reply:
For a query such as the one in the test, how do the context graph information flow?

```
MATCH (m) // no context graphs -- use ambient
WITH n // no context graphs still, as no graph return items -- this PR argues use ambient
RETURN n // no context graphs -- use ambient
```

The current model fails at the `WITH` -- there are no context graphs. This query should not fail, though. As I see it, there are two approaches:

a) Like this PR; no context graphs means ambient graph
b) Explicitly infer ambient graph as context graph if the leading clause is not a `WITH` with existing graph return items

The case you're describing is impossible to construct; there's no syntax that says 'project zero graphs', f.e. like `GRAPHS -`. You can however project two (non-context) graphs without explicitly setting new source and target; that's still an error and the other test in this PR makes sure that is still covered.

Reply:
Cool, thanks!